File: frontend/LocationTracker.js
/**
 * Location Tracker Utility
 * Real-time user location extraction from multiple sources
 * 
 * Usage:
 * - Start GPS tracking: LocationTracker.startGPSTracking()
 * - Update from Google Maps: LocationTracker.updateFromGoogleMaps(lat, lng, address)
 * - Update from geotag: LocationTracker.updateFromGeotag(imageFile)
 * - Get current location: LocationTracker.getCurrentLocation()
 */

class LocationTracker {
  constructor(backendUrl = "http://localhost:5000") {
    this.backendUrl = backendUrl;
    this.gpsWatchId = null;
    this.isTracking = false;
    this.currentLocation = null;
    this.updateInterval = 5000; // 5 seconds
    this.listeners = [];
  }

  /**
   * Start continuous GPS tracking from device
   * Called automatically on app start
   */
  startGPSTracking() {
    if (!navigator.geolocation) {
      console.error("Geolocation API not supported");
      return false;
    }

    this.isTracking = true;

    // High accuracy + continuous updates
    this.gpsWatchId = navigator.geolocation.watchPosition(
      (position) => this._handleGPSPosition(position),
      (error) => this._handleGPSError(error),
      {
        enableHighAccuracy: true,
        timeout: 10000,
        maximumAge: 0
      }
    );

    console.log("GPS tracking started");
    return true;
  }

  /**
   * Stop GPS tracking
   */
  stopGPSTracking() {
    if (this.gpsWatchId !== null) {
      navigator.geolocation.clearWatch(this.gpsWatchId);
      this.isTracking = false;
      console.log("GPS tracking stopped");
    }
  }

  /**
   * Handle GPS position update from device
   * Sends to backend continuously
   */
  _handleGPSPosition(position) {
    const { latitude, longitude, accuracy } = position.coords;

    // Send to backend
    this._sendToBackend("/location/track/gps", {
      latitude,
      longitude,
      accuracy
    });

    // Notify listeners
    this._notifyListeners({
      source: "gps",
      latitude,
      longitude,
      accuracy,
      timestamp: new Date().toISOString()
    });
  }

  /**
   * Handle GPS errors
   */
  _handleGPSError(error) {
    let message;
    switch (error.code) {
      case error.PERMISSION_DENIED:
        message = "Location permission denied. Please enable in settings.";
        break;
      case error.POSITION_UNAVAILABLE:
        message = "Location service unavailable";
        break;
      case error.TIMEOUT:
        message = "Location request timeout";
        break;
      default:
        message = "Unknown location error";
    }
    console.warn(message);
  }

  /**
   * Update location from Google Maps selection
   * Called when user clicks on map
   */
  updateFromGoogleMaps(latitude, longitude, address = "", placeId = null) {
    // Send to backend
    this._sendToBackend("/location/track/google-maps", {
      latitude,
      longitude,
      address,
      place_id: placeId
    });

    // Notify listeners
    this._notifyListeners({
      source: "google_maps",
      latitude,
      longitude,
      address,
      timestamp: new Date().toISOString()
    });
  }

  /**
   * Update location from image geotag
   * Extract GPS from uploaded image EXIF data
   */
  async updateFromGeotag(imageFile) {
    try {
      const formData = new FormData();
      formData.append("image", imageFile);

      const response = await fetch(`${this.backendUrl}/location/track/geotag`, {
        method: "POST",
        body: formData
      });

      const result = await response.json();

      if (result.success || response.status === 200) {
        const location = result.location || {};
        this._notifyListeners({
          source: "image_geotag",
          latitude: location.coordinates?.latitude,
          longitude: location.coordinates?.longitude,
          timestamp: new Date().toISOString()
        });
        return result;
      } else {
        throw new Error(result.error || "No geotag found in image");
      }
    } catch (error) {
      console.error("Geotag extraction error:", error);
      throw error;
    }
  }

  /**
   * Get current tracked location
   */
  async getCurrentLocation() {
    try {
      const response = await fetch(`${this.backendUrl}/location/current`);
      const result = await response.json();
      
      if (result.success) {
        this.currentLocation = result.location;
      }
      
      return result;
    } catch (error) {
      console.error("Error fetching current location:", error);
      return { success: false, error: error.message };
    }
  }

  /**
   * Get location tracking history
   */
  async getLocationHistory(limit = 10) {
    try {
      const response = await fetch(
        `${this.backendUrl}/location/history?limit=${limit}`
      );
      return await response.json();
    } catch (error) {
      console.error("Error fetching history:", error);
      return { error: error.message };
    }
  }

  /**
   * Get location tracking status
   */
  async getStatus() {
    try {
      const response = await fetch(`${this.backendUrl}/location/status`);
      return await response.json();
    } catch (error) {
      console.error("Error fetching status:", error);
      return { error: error.message };
    }
  }

  /**
   * Clear location tracking
   */
  async clearLocation() {
    try {
      const response = await fetch(`${this.backendUrl}/location/clear`, {
        method: "POST"
      });
      return await response.json();
    } catch (error) {
      console.error("Error clearing location:", error);
      return { error: error.message };
    }
  }

  /**
   * Subscribe to location updates
   */
  onLocationUpdate(callback) {
    this.listeners.push(callback);
  }

  /**
   * Notify all listeners of location change
   */
  _notifyListeners(locationData) {
    this.listeners.forEach((callback) => {
      try {
        callback(locationData);
      } catch (error) {
        console.error("Error in location listener:", error);
      }
    });
  }

  /**
   * Send data to backend
   */
  async _sendToBackend(endpoint, data) {
    try {
      const response = await fetch(`${this.backendUrl}${endpoint}`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(data)
      });

      if (!response.ok) {
        const error = await response.json();
        console.error(`Backend error at ${endpoint}:`, error);
      }
    } catch (error) {
      console.error(`Error sending to ${endpoint}:`, error);
    }
  }
}

// Export for use in React components
export default LocationTracker;

/**
 * USAGE IN REACT COMPONENT:
 * 
 * import LocationTracker from './LocationTracker';
 * 
 * useEffect(() => {
 *   const tracker = new LocationTracker();
 *   
 *   // Start continuous GPS tracking
 *   tracker.startGPSTracking();
 *   
 *   // Listen for location updates
 *   tracker.onLocationUpdate((location) => {
 *     console.log('Location updated:', location);
 *     setCurrentLocation(location);
 *   });
 *   
 *   return () => {
 *     tracker.stopGPSTracking();
 *   };
 * }, []);
 * 
 * // When user selects on Google Maps:
 * const handleMapClick = (lat, lng, address) => {
 *   tracker.updateFromGoogleMaps(lat, lng, address);
 * };
 * 
 * // When user uploads image:
 * const handleImageUpload = async (imageFile) => {
 *   try {
 *     const result = await tracker.updateFromGeotag(imageFile);
 *     console.log('Location extracted:', result);
 *   } catch (error) {
 *     console.error('No geotag in image');
 *   }
 * };
 */

File: backend/utils/location_tracker.py
"""
Real-time User Location Tracker
Continuously extracts and updates user location from multiple sources
"""

from geopy.geocoders import Nominatim
from utils.location_mapper import (
    format_location_for_complaint,
    validate_location_bounds,
    find_nearest_ward
)
import json
import time
from datetime import datetime


class LocationTracker:
    """
    Real-time location tracking with continuous updates.
    Supports: Browser Geolocation API, Google Maps, Image Geotag
    """
    
    def __init__(self):
        self.current_location = None
        self.location_history = []
        self.last_update = None
        self.update_interval = 5  # seconds
        
    
    def update_location_from_gps(self, latitude, longitude, accuracy=None):
        """
        Update location from GPS coordinates (real-time from device).
        Called continuously as user moves.
        
        Args:
            latitude: Current latitude
            longitude: Current longitude
            accuracy: GPS accuracy in meters (optional)
        
        Returns:
            dict: Updated location data
        """
        timestamp = datetime.now()
        
        # Validate bounds
        validation = validate_location_bounds(latitude, longitude)
        if not validation["valid"]:
            return {
                "success": False,
                "message": validation["message"]
            }
        
        # Format location
        location_data = format_location_for_complaint(latitude, longitude)
        location_data["source"] = "gps_continuous"
        location_data["timestamp"] = timestamp.isoformat()
        
        if accuracy:
            location_data["gps_accuracy_meters"] = accuracy
        
        # Update current location
        self.current_location = location_data
        self.last_update = timestamp
        
        # Track history
        self.location_history.append({
            "timestamp": timestamp.isoformat(),
            "latitude": latitude,
            "longitude": longitude,
            "accuracy": accuracy,
            "ward": location_data.get("ward", {}).get("ward_name")
        })
        
        return {
            "success": True,
            "location": location_data,
            "message": f"Location updated: {location_data.get('ward', {}).get('ward_name')}"
        }
    
    
    def update_location_from_google_maps(self, latitude, longitude, address, place_id=None):
        """
        Update location from Google Maps selection.
        User clicks on map and location is selected.
        
        Args:
            latitude: Selected latitude
            longitude: Selected longitude
            address: Human-readable address from Google Maps
            place_id: Google Maps place ID (optional)
        
        Returns:
            dict: Updated location data
        """
        timestamp = datetime.now()
        
        # Validate bounds
        validation = validate_location_bounds(latitude, longitude)
        if not validation["valid"]:
            return {
                "success": False,
                "message": validation["message"]
            }
        
        # Format location
        location_data = format_location_for_complaint(latitude, longitude)
        location_data["source"] = "google_maps"
        location_data["timestamp"] = timestamp.isoformat()
        location_data["address"]["google_address"] = address
        location_data["address"]["place_id"] = place_id
        
        # Update current location
        self.current_location = location_data
        self.last_update = timestamp
        
        # Track history
        self.location_history.append({
            "timestamp": timestamp.isoformat(),
            "latitude": latitude,
            "longitude": longitude,
            "address": address,
            "source": "google_maps"
        })
        
        return {
            "success": True,
            "location": location_data,
            "message": f"Location selected: {address}"
        }
    
    
    def update_location_from_geotag(self, geotag_data):
        """
        Update location from image geotag.
        
        Args:
            geotag_data: GPS data extracted from image EXIF
        
        Returns:
            dict: Updated location data
        """
        if not geotag_data:
            return {
                "success": False,
                "message": "No geotag data provided"
            }
        
        latitude = geotag_data.get("latitude")
        longitude = geotag_data.get("longitude")
        
        if not latitude or not longitude:
            return {
                "success": False,
                "message": "Invalid geotag coordinates"
            }
        
        timestamp = datetime.now()
        
        # Validate bounds
        validation = validate_location_bounds(latitude, longitude)
        if not validation["valid"]:
            return {
                "success": False,
                "message": validation["message"]
            }
        
        # Format location
        location_data = format_location_for_complaint(latitude, longitude)
        location_data["source"] = "image_geotag"
        location_data["timestamp"] = timestamp.isoformat()
        location_data["geotag_metadata"] = {
            "altitude": geotag_data.get("altitude"),
            "image_timestamp": geotag_data.get("timestamp"),
            "accuracy": geotag_data.get("accuracy_meters", 5)
        }
        
        # Update current location
        self.current_location = location_data
        self.last_update = timestamp
        
        # Track history
        self.location_history.append({
            "timestamp": timestamp.isoformat(),
            "latitude": latitude,
            "longitude": longitude,
            "source": "image_geotag",
            "ward": location_data.get("ward", {}).get("ward_name")
        })
        
        return {
            "success": True,
            "location": location_data,
            "message": f"Location extracted from image: {location_data.get('ward', {}).get('ward_name')}"
        }
    
    
    def get_current_location(self):
        """Get the most recent location."""
        if not self.current_location:
            return {
                "success": False,
                "message": "No location captured yet"
            }
        
        return {
            "success": True,
            "location": self.current_location,
            "last_update": self.last_update.isoformat() if self.last_update else None
        }
    
    
    def get_location_history(self, limit=10):
        """Get recent location history."""
        return {
            "history": self.location_history[-limit:],
            "total_updates": len(self.location_history)
        }
    
    
    def clear_location(self):
        """Clear current location data."""
        self.current_location = None
        self.location_history = []
        self.last_update = None
        return {"success": True, "message": "Location cleared"}


# Global tracker instance (shared across requests)
location_tracker = LocationTracker()
